### PR TITLE
feat: expand CIP4 with reference implementation

### PR DESCRIPTION
**Pull request by ramin**
## Overview

- expands reference implementation with details for sampling window for light nodes
- explains options for storage pruning
- updates CIP author
- updates link to PR for upcoming pruning PR in celestia-node in CIP preamble

### CODE REVIEW

File: cips/cip-4.md [hunk 1]
```diff
@@ -45,12 +45,43 @@ Data availability sampling light nodes SHOULD sample blocks created in the last
...
 2. Similarly, full nodes running on older software will not be able to sync historical blocks without discovering non-pruned nodes on the `archival` tag.
 3. Requesting blobs from historical blocks via a light node or full node will not be possible without discovering non-pruned nodes on the `archival` tag.
 
+## Reference Implementation
+
+### Data Availability Sampling Window (light nodes)
+
+Implementation for light nodes can be quite simple, where a satisfactory implementation merely behaves in that the choice to sample headers should not occur for headers whose timestamp is outside the given sampling window.
+
+Given a hypothetical "sample" function that performs data availability sampling of incoming extended headers from the network, the decision to sample or not should be taken by inspecting the header's timestamp, and ignoring it in any sampling operation if the duration between the header's timestamp and the current time exceeds the duration of the sampling window. For example
```
Comment (jcstein) [inline cips/cip-4.md:60]:
```suggestion
Given a hypothetical "sample" function that performs data availability sampling of incoming extended headers from the network, the decision to sample or not should be taken by inspecting the header's timestamp, and ignoring it in any sampling operation if the duration between the header's timestamp and the current time exceeds the duration of the sampling window. For example:
```

Comment (jcstein) [reply cips/cip-4.md:60]: [nit][not blocking]

File: cips/cip-4.md [hunk 2]
```diff
@@ -45,12 +45,43 @@ Data availability sampling light nodes SHOULD sample blocks created in the last
...
+
+```go
+
```
Comment (jcstein) [inline cips/cip-4.md:63]:
```suggestion
```

File: cips/cip-4.md [hunk 3]
```diff
@@ -45,12 +45,43 @@ Data availability sampling light nodes SHOULD sample blocks created in the last
...
+func sample(header Header) error{
+    if time.Since(header.Time()) > windowSize {
+        return nil // do not perform any sampling
+    }
+
+    // continue with rest of sampling operation
+}
+```
+
+Example implementation by [celestia node](https://github.com/celestiaorg/celestia-node/pull/2991)
```
Comment (jcstein) [inline cips/cip-4.md:75]:
```suggestion
[Example implementation by celestia node](https://github.com/celestiaorg/celestia-node/pull/2991)
```

Comment (jcstein) [reply cips/cip-4.md:75]: [nit][not blocking]

File: cips/cip-4.md [hunk 4]
```diff
@@ -45,12 +45,43 @@ Data availability sampling light nodes SHOULD sample blocks created in the last
...
+
+### Storage Pruning
+
+Pruning of data outside the availability window will be highly implementation specific and dependent on how data storage is engineered.
+
+A satisfactory implementation would be where any node implementing storage pruning may, if NOT advertising oneself to peers as an archival node on the 'full' topic, discard stored data outside the 30 day availability window. A variety of options exist for how any implementation might schedule pruning of data, and there are no requirements around how this is implemented. The only requirement is merely that the time guarantees around data within the availability window are properly respected, and that data availability nodes correctly advertise themselves to peers.
+
+An example implementation of storage pruning (WIP at time of writing) in [celestia node](https://github.com/celestiaorg/celestia-node/pull/3150/files)
```
Comment (jcstein) [inline cips/cip-4.md:83]:
```suggestion
[An example implementation of storage pruning (WIP at time of writing) in celestia node](https://github.com/celestiaorg/celestia-node/pull/3150/files)
```

**Review by jcstein — CHANGES_REQUESTED** [review]
LGTM, left minor styling edits

**Review by YazzyYaz — APPROVED** [review]

**Review by jcstein — APPROVED** [review]

**Pull request merged**